Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+
+While most Unicode-defined properties can only match at the Unicode scalar level, some are defined to match an extended grapheme cluster. For example, `/\p{RGI_Emoji_Flag_Sequence}/` will match any flag emoji character, which are composed of two Unicode scalar values.
+
+`\P{...}` matches the inverse of the characters matched by the corresponding `\p{...}` character class.
+
+**DSL:** This functionality is supported through APIs that take a predicate as a parameter, focused either on a `Character` or `UnicodeScalar` instance.
+
+```swift
+Pattern {
+    "abc"
+    OneOrMore(.unicodeProperty(where: \.isAlphabetic))
+    "def"
+}
+```
+
+### POSIX Character Classes: `[:NAME:]`
+
+The following POSIX character classes, of the form `[:NAME:]`, are supported:
+
+| POSIX class  | Matches           |
+|--------------|-------------------|
+| `[:alnum:]`  | `[A-Za-z0-9]`     |
+| `[:alpha:]`  | `[A-Za-z]`        |
+| `[:ascii:]`  | `[\x00-\x7F]`     |
+| `[:blank:]`  | `[ \t]`           |
+| `[:cntrl:]`  | `[\x00-\x1F\x7F]` |
+| `[:digit:]`  | `[0-9]`           |
+| `[:graph:]`  | `[\x21-\x7E]`     |
+| `[:lower:]`  | `[a-z]`           |
+| `[:print:]`  | `[[:graph:] ]`    |
+| `[:punct:]`  | `[-!"#$%&'()*+,./:;<=>?@[\\\]^_{|}~]` |
+| `[:space:]`  | `[ \t\n\r\f\v]`   |
+| `[:upper:]`  | `[A-Z]`           |
+| `[:word:]`   | `[A-Za-z0-9_]`    |
+| `[:xdigit:]` | `[0-9A-Fa-f]`     |
+
+### Custom Classes: `[...]`
+
+Users can create custom character classes by using literal characters and most predefined character classes within square brackets.
+
+- Individual characters and character classes described above can be used as is.
+- Ranges of Unicode scalar values can be specified by separating the start and end of the range (inclusive) with a hyphen. Note that because `Character` isn't a `Comparable` type, ranges can only match Unicode scalar values. Crossing the UTF-16 surrogate pair range (U+D800 to U+DFFF) is permitted, but values in that range do not match anything, regardless of the string's encoding.
+- Include a literal hyphen by escaping, or by placing it first or last in the custom class.
+- Negate a custom class by placing a caret immediately after the opening bracket (e.g. `[^a-zA-Z]`).
+
+
+## Alternatives Considered
+
+### Applying Regular Expressions to String Views
+
+A prior design allowed a user to choose their desired matching level and semantics by applying a regular expression to a string's `UnicodeScalarView`, `UTF16View`, or `UTF8View`. While initially appealing, this approach posed multiple problems:
+
+- Since we want regular expressions created with a literal to be usable all the same places as those created using the DSL, anything predicate-based needs to have the input type specified at the time of creation. It wouldn't make sense to add a `(Character) -> Bool` predicate to a regular expression, and then apply it to the string's `UTF8View`.
```

Comment:
How would you pass an arbitrary `(Character) -> Bool` for use within a regex literal anyways?